pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -859,6 +861,19 @@ func buildListener(opts buildListenerOpts) *xdsapi.Listener {
 	// TODO(incfly): consider changing this to map to handle duplicated listener filters from different chains?
 	var listenerFilters []listener.ListenerFilter
 
+	// add a TLS inspector if we need to detect ServerName or ALPN
+	needTLSInspector := false
+	for _, chain := range opts.filterChainOpts {
+		needsALPN := chain.tlsContext != nil && chain.tlsContext.CommonTlsContext != nil && len(chain.tlsContext.CommonTlsContext.AlpnProtocols) > 0
+		if len(chain.sniHosts) > 0 || needsALPN {
+			needTLSInspector = true
+			break
+		}
+	}
+	if needTLSInspector {
+		listenerFilters = append(listenerFilters, listener.ListenerFilter{Name: envoyListenerTLSInspector})
+	}
+
 	for _, chain := range opts.filterChainOpts {
 		listenerFilters = append(listenerFilters, chain.listenerFilters...)
```

Comment:
This won't help... 

Just reading the code, you just optionally add `envoyListenerTLSInspector`, which only add cases where tls inspector are added to sidecar config. `listenerFilters` slice now can only have potential more filters to add, but not less.

Reply:
Could you expand a bit? I don't think I follow.

Reply:
So the `listenerFilters` holds the listener level filters. Without this PR, in your case #7601, sidecar receives `tls_inspector` somehow. With your PR, it's only possible to make `listenerFilters` contain more elements. The `tls_inspector` added somewhere else (through line 878) will still be in the array.

Reply:
Oh, sorry, re-read the issue, the problem is the opposite: we need `tls_inspector` but it's not there?

Reply:
Yeah, we need TLS inspector but it isn't there. Right now Envoy is saving us by automatically adding it, but it spits out a warning message. I'm trying to add the TLS inspector so that Envoy does not have to add it and write a warning.

Reply:
@incfly: does that make sense? Would you be willing to remove the hold?

Reply:
Sorry, I'm OOO right now and reply is spotty...

Yeah, that definitely makes sense. Just one minor thing

The `Plugin.OnFilterChains` might add listener filters(`tls_inspector` might already be included as a result of Line 878. ). Not sure if it's harmless would be better to avoid duplicates if possible.

So can you make it such append `tls_inspector` only if `listenerFilters` does not have it?

/cc @rshriram @PiotrSikora

Reply:
Makes sense. We don't add it right now, but that could change. I'll update the logic in another PR, since this one merged (and nothing is broken at the moment).